### ISSUE TITLE
Handle application exceptions with 500 errors

### ISSUE THREAD

**Issue by fizruk** [haskell-servant/servant#309]
In https://github.com/yesodweb/wai/issues/496#issuecomment-167382191 @snoyberg suggests that application errors should not get to the logger middleware and rather get caught by the application itself.

#303 is also related.

I do not have a good understanding of where the handling should happen so I am filing this issue to at least open a discussion.

**Comment by jkarni** [comment]
Aha, now I see why #303 was happening. And @snoyberg 's suggestion makes sense - `warp` can't really catch exceptions before they hit the middleware. 

[toApplication](https://github.com/haskell-servant/servant/blob/6a4c96759053a69d75514bf356291575c9cd5ad0/servant-server/src/Servant/Server/Internal/RoutingApplication.hs#L40) seems like a reasonable place to do it. I guess we might have to force the response to discover latent errors, but that doesn't sound great.

**Comment by snoyberg** [comment]
Yesod by default fully evaluated response bodies, and offers the ability to
explicitly annotate a response as not requiring full evaluation.

**Comment by jkarni** [comment]
@snoyberg thanks! That sounds like a pretty good solution.

**Comment by fizruk** [comment]
cc @mwotton

**Comment by mwotton** [comment]
cheers all for the lucid analysis & discussion :) after thinking about it a bit more I understand the objection to doing it in the middleware.

**Comment by v0d1ch** [comment]
Ok so this is marked as newcomer-friendly and I get to ask questions then :) 
If this is the place to handle 500 -> [toApplication](https://github.com/haskell-servant/servant/blob/6a4c96759053a69d75514bf356291575c9cd5ad0/servant-server/src/Servant/Server/Internal/RoutingApplication.hs#L40 ) do I need to throw error from this function, something like this `throwError $ err500 { errBody = "Oooops" }` or am I missing the point

**Comment by v0d1ch** [comment]
Ok this is marked as newcomer-friendly so I get to ask questions :)
If [this](https://github.com/haskell-servant/servant/blob/6a4c96759053a69d75514bf356291575c9cd5ad0/servant-server/src/Servant/Server/Internal/RoutingApplication.hs#L40) is a place to add changes can I throw 500 error if : 

`routingRespond (FailFatal err)` 

Something like this
 
`throwError $ err500 { errBody = "Oooops!" }`

**Comment by jkarni** [comment]
The file changed somewhat - latest version is [here](https://github.com/haskell-servant/servant/blob/master/servant-server/src/Servant/Server/Internal/RoutingApplication.hs#L92).

My intuition is: `toApplication` should take an extra argument, representing whether to fully evaluate the response or not. If so, we need to `seq` or `nf` the response. (If not, things stay as they are). Something like 

```haskell
toApplication :: Bool -> RoutingApplication -> Application
toApplication fullyEvaluate ra request respond = ra request (maybeEval . routingRespond)
 where
  maybeEval x = if fullyEvaluate then nf <$> x else x
  routingRespond :: RouteResult Response -> IO ResponseReceived
  routingRespond (Fail err)      = respond $ responseServantErr err
  routingRespond (FailFatal err) = respond $ responseServantErr err
routingRespond (Route v) = respond v
```

Then we should use `Context` to determine whether to fully evaluate (default = no). That way it's possible to configure this globally or for sub-APIs.

**Comment by v0d1ch** [comment]
I would need some guidance on this since I am still a noob or somebody else can do this until I get the hang of it.

**Comment by alpmestan** [comment]
Oh no need to go in private. Please ask your questions here and other people will do their best to guide you towards a good solution :)

**Comment by v0d1ch** [comment]
Ok Ill give it a shot, mind you I am over my head here...
```
maybeEval x = if fullyEvaluate then nf <$> x else x
```
So if we don't want to fully evaluate response (which is default) we just return routingRespond var and else we 'seq or nf' ?
So seq is most probably [this](https://wiki.haskell.org/Seq) 
`seq :: a -> b -> b` 
and what about nf ? [this](https://hackage.haskell.org/package/nf) one ? 
do I need to import it since i don't see it is loaded ?

so how do I get normal form of the response ? what is the second param for seq if first one is response, do we throw 500 there ?  and I am also confused regarding where Context part go? So basically everything confuses me :)

**Comment by v0d1ch** [comment]
@alpmestan How about a little clue :)

**Comment by alpmestan** [comment]
nf is actually probably [`rnf`](https://hackage.haskell.org/package/deepseq-1.4.2.0/docs/Control-DeepSeq.html#v:rnf), and `seq` is indeed the one you point to.

I don't think we want to force the evaluation of the `ResponseReceived` value, but instead of the `Response` (which doesn't have an `NFData` instance either). @jkarni thoughts?

**Comment by v0d1ch** [comment]
I managed to get it compiled with deepseq dependency. I hardcoded the Bool param in Server.hs for now:
```
serveWithContext :: (HasServer api context)
    => Proxy api -> Context context -> Server api -> Application
serveWithContext p context server =
  toApplication True (runRouter (route p context (emptyDelayed (Route server))))
```
I am having some problems with NFData instance for Response since I don't think this is correct 
```
-- Internal.hs :
instance NFData Response where
    rnf a = seq a ()
```
and I tried to nf Response like this 
```
toApplication :: Bool -> RoutingApplication -> Application
toApplication fullyEvaluate ra request respond = ra request routingRespond
 where
  maybeEval x = if fullyEvaluate then rnf x  else  x
  routingRespond :: RouteResult Response -> IO ResponseReceived
  routingRespond (Fail err)      = respond $ maybeEval $ responseServantErr err
  routingRespond (FailFatal err) = respond $ maybeEval $ responseServantErr err
  routingRespond (Route v)       =  respond $ maybeEval v
```
But it throws 
```
ghci > :r
[3 of 3] Compiling Servant.Server.Internal.RoutingApplication ( servant-server/src/Servant/Serve
r/Internal/RoutingApplication.hs, interpreted )                                                

servant-server/src/Servant/Server/Internal/RoutingApplication.hs:98:46:
    Couldn't match expected type ‘Response’ with actual type ‘()’
    In the second argument of ‘($)’, namely
      ‘maybeEval $ responseServantErr err’
    In the expression: respond $ maybeEval $ responseServantErr err
    In an equation for ‘routingRespond’:
        routingRespond (Fail err)
          = respond $ maybeEval $ responseServantErr err

servant-server/src/Servant/Server/Internal/RoutingApplication.hs:98:58:
    Couldn't match expected type ‘()’ with actual type ‘Response’
    In the second argument of ‘($)’, namely ‘responseServantErr err’
    In the second argument of ‘($)’, namely
      ‘maybeEval $ responseServantErr err’

servant-server/src/Servant/Server/Internal/RoutingApplication.hs:99:46:
    Couldn't match expected type ‘Response’ with actual type ‘()’
    In the second argument of ‘($)’, namely
      ‘maybeEval $ responseServantErr err’
    In the expression: respond $ maybeEval $ responseServantErr err
    In an equation for ‘routingRespond’:
        routingRespond (FailFatal err)
          = respond $ maybeEval $ responseServantErr err

servant-server/src/Servant/Server/Internal/RoutingApplication.hs:99:58:
    Couldn't match expected type ‘()’ with actual type ‘Response’
    In the second argument of ‘($)’, namely ‘responseServantErr err’
    In the second argument of ‘($)’, namely
      ‘maybeEval $ responseServantErr err’

servant-server/src/Servant/Server/Internal/RoutingApplication.hs:100:41:
    Couldn't match expected type ‘Response’ with actual type ‘()’
    In the second argument of ‘($)’, namely ‘maybeEval v’
    In the expression: respond $ maybeEval v

servant-server/src/Servant/Server/Internal/RoutingApplication.hs:100:51:
    Couldn't match expected type ‘()’ with actual type ‘Response’
    In the first argument of ‘maybeEval’, namely ‘v’
    In the second argument of ‘($)’, namely ‘maybeEval v’
Failed, modules loaded: Servant.Server.Internal.Handler, Servant.Server.Internal.ServantErr.

```
So basically I don't know how to construct correct Response from this line since rnf x returns unit () 
`maybeEval x = if fullyEvaluate then rnf x  else  x`

**Comment by v0d1ch** [comment]
@jkarni what are your thoughts on this ?

**Comment by rfn01** [comment]
Hey, I might be able to help just a little, and would find this feature useful.

Looks like `rnf` returns `()`.  You might want `force :: NFData a => a -> a` instead.

The `Control.DeepSeq` doc under `rnf` says we can get `NFData` derived for us automatically with recent GHC and a couple of extensions.  We need this for `RouteResult` and `ServantErr`, that are in `servant-server/src/Servant/Server/Internal/RoutingApplication.hs` and `servant-server/src/Servant/Server/Internal/ServantErr.hs`.  To both files add these extensions and imports:

```
{-# LANGUAGE DeriveGeneric              #-}
{-# LANGUAGE DeriveAnyClass             #-}

import GHC.Generics (Generic)
import Control.DeepSeq
```
and then add `Generic` and `NFData` to the deriving clause for those two types.

(I'm a noob, so please tell me if I'm wrong, or did something wrong, or really anything you'd like to tell me.)

**Comment by v0d1ch** [comment]
Hi @rfn01 , thanks for joining! Did you get this to compile ? I derived instances like you suggested and removed my NFData instance for Response datatype.

I get 
```
...
couldn't match ResponseReceived with ()
Expected type RouteResult Response -> IO ()
Actual type RouteResult Response -> IO ResponseReceived
...

 -- RoutingApplication.hs
toApplication :: Bool -> RoutingApplication -> Application
toApplication fullyEvaluate ra request respond = ra request (maybeEval . routingRespond)
 where
  maybeEval x = if fullyEvaluate then rnf <$> x else  x
  routingRespond :: RouteResult Response -> IO ResponseReceived
  routingRespond (Fail err)      = respond $ responseServantErr err
  routingRespond (FailFatal err) = respond $ responseServantErr err
  routingRespond (Route v) = respond v

```

**Comment by rfn01** [comment]
Try using `force` instead of `rnf`, sort of like this (where I've skipped checking `fullyEvaluate` for now):

```
toApplication ra request respond = ra request (force routingRespond)
```

**Comment by v0d1ch** [comment]
Right this compiles since it does not have to return unit. I was following blindly to what @jkarni has wrote and tried stubbornly to use rnf. I guess since both functions perform essentially the same evaluation this should be fine. Thanks!

**Comment by v0d1ch** [comment]
This leaves the question of extracting the Bool value from Context.
I learned the hard way that newcomer-friendly is not the same as beginner-friendly :)

**Comment by rfn01** [comment]
Hey @v0d1ch,

`force` is just a handy way to get `rnf` called in this situation.  It's the same.  Take a peek at the source (maybe via the "source" links in the [documentation](https://hackage.haskell.org/package/deepseq-1.4.2.0/docs/Control-DeepSeq.html)), and you'll find something that is sort of like this:

```
force x = x `deepseq` x
deepseq a b = rnf a `seq` b
```

I haven't used Context before.  Give it a try and ask questions.

/rob

**Comment by v0d1ch** [comment]
Yes I saw that, thanks. I'll play with Context and see if I can nail it, I never worked with DataKinds extension before.

**Comment by qrilka** [comment]
@v0d1ch did you continue to work on this one?

**Comment by v0d1ch** [comment]
Hey @qrilka I kinda completely forgot about this. If you want you can continue otherwise I will do it.

**Comment by qrilka** [comment]
Sorry, not sure that I'll have enough time for it at the moment

**Comment by v0d1ch** [comment]
Ok then nice of you to remind me about this, I'll try to put something together

**Comment by domenkozar** [comment]
Maybe servant should just document https://hackage.haskell.org/package/warp-3.2.23/docs/Network-Wai-Handler-Warp.html#v:setOnException and provide a default handler?

**Comment by WraithM** [comment]
`setOnException` doesn't work for `throwError` in a `Handler`. It seems that if you throw an exception in IO (ie. `liftIO $ throw someException`), that gets handled by the default exception handler in warp. However, throwing a `ServantErr` doesn't get caught by warp. Servant responds with a 500 error, but nothing gets logged.

**Comment by jeremyjh** [comment]
@domenkozar how does `setOnException` help us with the original problem? If the exception makes it all the way to the Warp handler, it will by default send a 500 response. What it will not do, is go back and tell the logging middleware about that response. The exception needs to be handled before any middleware are applied.

**Comment by alpmestan** [comment]
@wraithm `setOnException` is only used for exceptions. `throwError` on the other hand will just trigger the use of the `Left` branch of some `Either` somewhere in your monad stack, to shortcircuit everything else. These are two different mechanisms, one which has some special support from the runtime system (exceptions), the other which just amounts to doing `Left "user foo@bar.com doesn't exist"` and using `Either`'s r `ExceptT`'s Monad instance to prevent the rest of the code from being executed.

Exceptions are "implicit" (they don't appear in types) and just bubble up until they're caught. Servant doesn't currently catch any of them and just lets warp's exception handler do the work. You can use [`setOnException`](https://hackage.haskell.org/package/warp-3.2.23/docs/Network-Wai-Handler-Warp.html#v:setOnException) to change the default handler for exceptions. Now, it's arguably even better to `catch` the exception yourself, from your handler, and directly generate a nice, helpful response using all the information you've got available there, instead of trying to recover that later down the road.

We can of course still discuss whether servant itself could/should catch some/all exceptions raised while running the request handlers that the user supplied. I personally don't think so, since we would not have any more useful information than warp.

tl;dr: If your handlers use functions that can throw exceptions, I'd definitely recommend grabbing them from your handlers directly. Other people have other preferences, this is just my opinion.

**Comment by jeremyjh** [comment]
@alpmestan if you let warp catch the error, none of the middlewares will run. This means if you have installed Wai logging middleware, it will not log 500 responses when your application throws an exception. Log monitors that you've setup to report on things like 500 errors will cheerfully inform your dashboards that everything is fine. Personally, I consider this to be an unacceptable default behavior from a web application framework.

**Comment by alpmestan** [comment]
I don't like that behaviour either, and that's precisely why I never let my exceptions make it that far. I'm better equipped for dealing with them in my handlers, where I've got my logging, error reporting, etc machinery in place and available to me.

What would you like servant-server to do, in the situation you described? Wrap all handlers with a "catch-all" block that would prevent exceptions from ever hitting warp? Wouldn't we want that to apply to all wai/warp apps, whether built with servant or not?

**Comment by jeremyjh** [comment]
@alpmestan I agree Handlers are better equipped to deal with known exceptions, but what pretty much every other web framework does, is just catch all uncaught exceptions from application code, log the exception and respond with a generic 500 error. Warp does do this, but it is too late since the exception has already bubbled through all the middlware. This needs to be handled in the bottom layer application before middleware are applied. This is what Yesod does, and is what the author of Yesod and Wai recommends other web frameworks do.

**Comment by alpmestan** [comment]
Right, but I'm not a huge fan of choices that we cannot opt out of, and I think I'm not alone. For instance, in _all_ of my applications, this would be redundant and might even waste cycles (not a lot, of course, but the fact that it would waste them for something so trivial bothers me). I have also seen many open-source and closed-source apps doing the same thing as me.

The best option probably is leave the choice to the user. I'm not sure which form that could take, but it would also encourage users to think about what exceptions they might deal with and what they want to do with them. Just a nice side-effect.

Also, with a slightly different definition of `Middleware`, we could just "go and insert" the exception catching block from warp at the innermost layer, just around the actual app, once and for all, for all the web frameworks that can target it. This is IMO the right solution (as exceptions otherwise basically break the "middleware promise" so to speak), but this is hard to change at this point I suppose...

**Comment by k-bx** [comment]
> `setOnException` doesn't work for `throwError` in a `Handler`. It seems that if you throw an exception in IO (ie. `liftIO $ throw someException`), that gets handled by the default exception handler in warp. However, throwing a `ServantErr` doesn't get caught by warp. Servant responds with a 500 error, but nothing gets logged.
> 
> My opinion is that this is counter-intuitive. I'd expect any 5xx type error to be logged, regardless if it's a `ServantErr` that gets thrown by `throwError` or an exception in IO.

Please see https://github.com/yesodweb/wai/pull/770 , should be fixed there.

**Comment by phadej** [comment]
I’m inclined to remove `Exception ServerError` instance. It’s not intended to be thrown.

**Comment by alpmestan** [comment]
I'd approve that.

**Comment by wraithm** [comment]
I use the `Exception` instance for 301 redirects. Maybe that's an incorrect use of `ServerError`, but I can't figure out another way to do it.

**Comment by alpmestan** [comment]
You can use the `MonadExcept` instance, which is compatible with anything `Either`-ish, like `Handler` (which uses `ExceptT ServerError`). `throwError (err404 { ... })` should just work. An alternative is to use your own monad with your own domain-specific error type, and then perform the conversion to HTTP errors in the transformation from `YourMonad` to `Handler`.

**Comment by wraithm** [comment]
Great. I'm sure that I'm actually using the `MonadExcept` instance. I just got confused.
